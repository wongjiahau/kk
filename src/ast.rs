use crate::non_empty::NonEmpty;
/// The syntax tree here represents raw syntax tree that is not type checked

#[derive(Debug, Clone)]
pub enum Statement {
    Let {
        keyword_let: Token,
        left: Token,
        type_variables: Vec<Token>,
        right: Expression,
        type_annotation: Option<TypeAnnotation>,
    },
    Type {
        keyword_type: Token,
        left: Token,
        right: TypeAnnotation,
        type_variables: Vec<Token>,
    },
    Enum {
        keyword_enum: Token,
        name: Token,
        constructors: Vec<EnumConstructor>,
        type_variables: Vec<Token>,
    },
    Do {
        keyword_do: Token,
        expression: Expression,
    },
}

#[derive(Debug, Clone)]
pub struct EnumConstructor {
    pub name: Token,
    pub left_parenthesis: Token,
    pub payload: Option<Box<TypeAnnotation>>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct TypeVariable {
    pub token: Token,
}

#[derive(Debug, Clone)]
pub enum Type {
    Underscore,

    /// Type variable that is declared. Cannot be substituted before instantiation.
    /// Note that a type variable is only explicit within its own scope
    ExplicitTypeVariable {
        name: String,
    },

    /// Type variable that is implicitly created  for unification.
    ImplicitTypeVariable {
        name: String,
    },
    Record {
        key_type_pairs: Vec<(String, Type)>,
    },
    Named {
        name: String,
        type_arguments: Vec<(String, Type)>,
    },
    Function(FunctionType),
    Tuple(Box<NonEmpty<Type>>),
    Boolean,
    Float,
    Integer,
    String,
    Character,
    Null,
    Array(Box<Type>),
}

#[derive(Debug, Clone)]
pub struct TagType {
    pub tagname: String,
    pub payload: Option<Box<Type>>,
}

#[derive(Debug, Clone)]
pub struct TypeScheme {
    pub type_variables: Vec<String>,
    pub type_value: Type,
}

#[derive(Debug, Clone)]
pub struct FunctionType {
    pub parameters_types: Box<NonEmpty<Type>>,
    pub return_type: Box<Type>,
}

#[derive(Debug, Clone)]
pub enum SymbolSource {
    BuiltIn,
    UserDefined {
        filename: String,
        position: Position,
    },
}

#[derive(Debug, Clone)]
pub struct NamedTypeAnnotationArguments {
    pub left_angular_bracket: Token,
    pub arguments: Vec<TypeAnnotation>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub enum TypeAnnotation {
    Named {
        name: Token,
        type_arguments: Option<TypeArguments>,
    },
    Record {
        left_curly_bracket: Token,
        key_type_annotation_pairs: Vec<(Token, TypeAnnotation)>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        element_type: Box<TypeAnnotation>,
        right_square_bracket: Token,
    },
    Underscore(Token),
    Function {
        start_token: Token,
        parameters_types: Box<NonEmpty<TypeAnnotation>>,
        return_type: Box<TypeAnnotation>,
    },
}

#[derive(Debug, Clone)]
pub enum DestructurePattern {
    String(Token),
    Character(Token),
    Integer(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Null(Token),
    Underscore(Token),
    Identifier(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<DestructurePattern>>,
        right_parenthesis: Token,
    },
    Record {
        left_curly_bracket: Token,
        key_value_pairs: Vec<DestructuredRecordKeyValue>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        right_square_bracket: Token,
        spread: Option<DestructurePatternArraySpread>,
    },
    Tuple(DestructurePatternTuple),
}
#[derive(Debug, Clone)]
pub struct DestructurePatternTuple {
    pub parentheses: Option<(/*left*/ Token, /*right*/ Token)>,
    pub values: Box<NonEmpty<DestructurePattern>>,
}

#[derive(Debug, Clone)]
pub struct DestructurePatternArraySpread {
    pub first_element: Box<DestructurePattern>,
    pub spread_token: Token,
    pub rest_elements: Box<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub struct DestructuredRecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub as_value: Option<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub enum Expression {
    Null(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Float(Token),
    Integer(Token),
    String(Token),
    Character(Token),
    Variable(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<Expression>>,
        right_parenthesis: Token,
    },
    Function(Box<Function>),
    FunctionCall(Box<FunctionCall>),
    Record {
        left_curly_bracket: Token,
        spread: Option<Box<Expression>>,
        key_value_pairs: Vec<RecordKeyValue>,
        right_curly_bracket: Token,
    },
    RecordAccess {
        expression: Box<Expression>,
        property_name: Token,
    },
    Array {
        left_square_bracket: Token,
        elements: Vec<Expression>,
        right_square_bracket: Token,
    },
    Let {
        keyword_let: Token,
        left: Box<DestructurePattern>,
        type_annotation: Option<TypeAnnotation>,
        right: Box<Expression>,
        true_branch: Box<Expression>,
        false_branch: Option<Box<Function>>,
    },
}

#[derive(Debug, Clone)]
pub struct RecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub value: Expression,
}

#[derive(Debug, Clone)]
pub struct FunctionCall {
    pub function_name: Token,
    pub first_argument: Box<Expression>,
    pub rest_arguments: Option<FunctionCallRestArguments>,
    pub type_arguments: Option<TypeArguments>,
}

#[derive(Debug, Clone)]
pub struct TypeArguments {
    pub left_angular_bracket: Token,
    pub substitutions: Vec<(/*type_variable_name*/ Token, TypeAnnotation)>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionCallRestArguments {
    pub left_parenthesis: Token,
    pub arguments: Vec<Expression>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct Function {
    pub branches: NonEmpty<FunctionBranch>,
}

#[derive(Debug, Clone)]
pub struct FunctionBranch {
    pub start_token: Token,
    pub parameters: FunctionParameters,
    pub body: Box<Expression>,
    pub return_type_annotation: Option<TypeAnnotation>,
}

impl FunctionBranch {
    pub fn parameters(&self) -> NonEmpty<FunctionParameter> {
        match &self.parameters {
            FunctionParameters::NoParenthesis { parameter } => NonEmpty {
                head: parameter.clone(),
                tail: vec![],
            },
            FunctionParameters::WithParenthesis { parameters, .. } => parameters.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum FunctionParameters {
    NoParenthesis {
        parameter: FunctionParameter,
    },
    WithParenthesis {
        left_parenthesis: Token,
        parameters: NonEmpty<FunctionParameter>,
        right_parenthesis: Token,
    },
}

#[derive(Debug, Clone)]
pub struct FunctionBranchRestArguments {
    pub left_parenthesis: Token,
    pub rest_arguments: Vec<FunctionParameter>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionParameter {
    pub destructure_pattern: DestructurePattern,
    pub type_annotation: Option<TypeAnnotation>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub position: Position,
    pub representation: String,
}

#[derive(Debug, Clone)]
pub struct ScopedName {
    pub namespaces: Vec<Token>,
    pub name: Token,
}

impl Token {
    pub fn dummy_identifier(representation: String) -> Token {
        Token {
            token_type: TokenType::Identifier,
            position: Position {
                line_start: 0,
                line_end: 0,
                character_index_start: 0,
                character_index_end: 0,
                column_start: 0,
                column_end: 0,
            },
            representation,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenType {
    KeywordLet,
    KeywordType,
    KeywordEnum,
    KeywordDo,
    KeywordElse,
    KeywordNull,
    KeywordTrue,
    KeywordFalse,
    Whitespace,
    LeftCurlyBracket,
    RightCurlyBracket,
    LeftParenthesis,
    RightParenthesis,
    LeftSquareBracket,
    RightSquareBracket,
    Newline,
    Colon,
    LessThan,
    MoreThan,
    Equals,
    Period,
    Spread,
    Comma,
    Minus,
    FatArrowRight,
    ThinArrowRight,
    Backslash,
    Underscore,
    Identifier,
    String,
    Character,
    Integer,
    Float,
    ScopeResolution,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Source {
    File { path: String },
    NonFile { env_name: String },
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct Position {
    pub line_start: usize,
    pub line_end: usize,
    pub column_start: usize,
    pub column_end: usize,
    pub character_index_start: usize,
    pub character_index_end: usize,
}
