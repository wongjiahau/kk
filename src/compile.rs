use crate::ast::*;
use crate::parse::parse_statements;
use crate::stringify_error::print_unify_error;
use crate::tokenize::tokenize;
use crate::transpile::transpile_statements;
use crate::unify::{unify_program, Program};

pub fn compile(source: Source, code: String) {
    match source_to_statements(code.clone()) {
        Err(parse_error) => {
            println!("{:#?}", parse_error);
        }
        Ok(statements) => match unify_program(Program {
            source: source.clone(),
            statements: statements.clone(),
        }) {
            Err(unify_error) => print_unify_error(source, code, unify_error),
            Ok(_) => {
                use std::process::Command;
                let javascript = transpile_statements(statements);
                let output = Command::new("node")
                    .arg("-e")
                    .arg(javascript)
                    .output()
                    .expect("Failed to run NodeJS binary");

                let stdout = String::from_utf8_lossy(&output.stdout).to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).to_string();
                if !stdout.is_empty() {
                    println!("{}", stdout.trim())
                }
                if !stderr.is_empty() {
                    eprintln!("{}", stderr.trim())
                }
            }
        },
    }
}

pub fn source_to_statements(source: String) -> Result<Vec<Statement>, ParseError> {
    let tokens = tokenize(source)?;
    parse_statements(tokens)
}
