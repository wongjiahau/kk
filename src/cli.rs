use crate::compile::compile;
use crate::{ast::Source, unify::Program};
use clap::Clap;
use std::fs;

#[derive(Clap)]
struct Opts {
    #[clap(subcommand)]
    subcmd: SubCommand,
}

#[derive(Clap)]
enum SubCommand {
    /// Execute a KK script
    Run(Run),
}

#[derive(Clap)]
struct Run {
    /// Input filename
    filename: String,
}

pub fn cli() {
    let opts: Opts = Opts::parse();

    match opts.subcmd {
        SubCommand::Run(run) => {
            match fs::read_to_string(&run.filename) {
                Err(_) => {
                    eprintln!("Unable to find file '{}'", run.filename);
                }
                Ok(code) => compile(Program {
                    source: Source::File { path: run.filename },
                    code,
                    import_relations: vec![], // empty, because this is the root
                }),
            }
        }
    }
}
