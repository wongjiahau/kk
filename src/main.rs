pub mod ast;
use ast::*;

mod environment;

mod unify;
use unify::*;

mod transpile;
use transpile::*;

mod parse;

mod tokenize;
use tokenize::*;

mod stringify_error;
use stringify_error::*;

mod cli;
use cli::*;

mod compile;
use compile::*;

mod pattern;

#[test]
fn run_all_tests() {
    use colored::*;
    use insta::assert_snapshot;
    use std::fs;
    use std::process::Command;

    let test_dir = "tests/compiler/";
    let folders = fs::read_dir(test_dir).expect("Failed to read directory");
    for folder in folders {
        let files = fs::read_dir(folder.unwrap().path().to_str().unwrap())
            .expect("Failed to read directory");
        for file in files {
            let file = file.expect("Failed to read entry");
            let filename = file
                .path()
                .to_str()
                .expect("Failed to convert entry to string")
                .to_string();

            if filename.ends_with(".kk") {
                let input_filename = filename;
                let input = fs::read_to_string(&input_filename).expect("failed to read input file");
                print!("{}", input_filename);
                let actual_output = {
                    let output = Command::new("./target/debug/kk")
                        .arg("run")
                        .arg(&input_filename)
                        .output()
                        .expect("Failed to run KK CLI");

                    let exit_code = match output.status.code() {
                        Some(code) => code.to_string(),
                        None => "".to_string(),
                    };

                    vec![
                        "============".to_string(),
                        "INPUT".to_string(),
                        "============".to_string(),
                        input.trim().to_string(),
                        "".to_string(),
                        "============".to_string(),
                        "EXIT CODE".to_string(),
                        "============".to_string(),
                        exit_code,
                        "".to_string(),
                        "============".to_string(),
                        "STDOUT".to_string(),
                        "============".to_string(),
                        strip_line_trailing_spaces(
                            String::from_utf8_lossy(&output.stdout.clone()).to_string(),
                        )
                        .trim()
                        .to_string(),
                        "".to_string(),
                        "============".to_string(),
                        "STDERR".to_string(),
                        "============".to_string(),
                        strip_line_trailing_spaces(
                            String::from_utf8_lossy(&output.stderr.clone()).to_string(),
                        )
                        .trim()
                        .to_string(),
                    ]
                    .join("\n")
                };

                let actual_output = strip_line_trailing_spaces(actual_output);

                let stripped_actual_output = strip_line_trailing_spaces(
                    String::from_utf8(
                        strip_ansi_escapes::strip(actual_output.clone())
                            .expect("Failed to strip color"),
                    )
                    .unwrap(),
                );

                assert_snapshot!(input_filename.clone(), stripped_actual_output.trim());
                println!("{}", " PASSED".green());

                // if stripped_actual_output.trim() != expected_output {
                //     let changeset = Changeset::new(&expected_output, &stripped_actual_output, "");
                //     println!("{}", "=".repeat(10));
                //     println!(
                //         "ASSERTION FAILED FOR:\n\n{}",
                //         indent_string(input_filename.to_string(), 4)
                //     );
                //     println!("\n\nINPUT:\n\n{}", indent_string(input, 4));
                //     println!(
                //         "\n\nEXPECTED OUTPUT:\n\n{}",
                //         indent_string(expected_output, 4)
                //     );
                //     println!(
                //         "\n\nACTUAL OUTPUT({}):\n\n{}",
                //         output_filename,
                //         indent_string(actual_output, 4)
                //     );
                //     println!(
                //         "\n\nDIFF (EXPECTED OUTPUT / ACTUAL OUTPUT):\n\n{}",
                //         indent_string(changeset.to_string(), 4)
                //     );
                //     println!("{}", "=".repeat(10));
                //     panic!("ASSERTION FAILED.")
                // } else {
                //     println!("{}", " PASSED".green());
                // }
            }
        }
    }

    fn strip_line_trailing_spaces(input: String) -> String {
        input
            .split('\n')
            .into_iter()
            .map(|line| line.trim_end().to_string())
            .collect::<Vec<String>>()
            .join("\n")
    }
}

fn main() {
    cli();
}

#[derive(Debug)]
pub enum CompileError {
    UnifyError(UnifyError),
    ParseError(ParseError),
}

pub fn type_check_source(code: String) -> String {
    let source = Source::NonFile {
        env_name: "TEST".to_string(),
    };
    compile(source, code);
    panic!()
}

pub fn transpile_source(source: String) -> Result<String, ParseError> {
    let statements = source_to_statements(source)?;
    // println!("{:#?}", statements);
    Ok(transpile_statements(statements))
}

pub fn execute(source: String) -> String {
    let statements = source_to_statements(source).unwrap();
    // println!("{:#?}", statements);
    let javascript = transpile_statements(statements);
    use std::process::Command;
    let output = Command::new("node")
        .arg(format!("-e \"{}\"", javascript))
        .output()
        .expect("failed to execute process");
    vec![
        format!("status = {}", output.status),
        format!("stdout = {}", String::from_utf8_lossy(&output.stdout)),
        format!("stderr = {}", String::from_utf8_lossy(&output.stderr)),
    ]
    .join("\n")
}

#[cfg(test)]
mod test_tokenize {
    use super::*;

    #[test]
    fn test_tokenize_arrow_right() {
        assert_eq!(
            tokenize("=>".to_string()),
            Ok(vec![Token {
                token_type: TokenType::ArrowRight,
                representation: "=>".to_string(),
                position: Position {
                    column_start: 0,
                    column_end: 1,
                    line_start: 0,
                    line_end: 0,
                    character_index_start: 0,
                    character_index_end: 0
                }
            }])
        );
    }

    #[test]
    fn test_tokenize_tag() {
        assert_eq!(
            tokenize("#Hello_world123 ".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::Tag,
                    representation: "#Hello_world123".to_string(),
                    position: Position {
                        column_start: 0,
                        column_end: 14,
                        line_start: 0,
                        line_end: 0,
                        character_index_start: 0,
                        character_index_end: 0
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 15,
                        column_end: 15,
                        character_index_start: 0,
                        character_index_end: 0
                    }
                }
            ])
        );
    }

    #[test]
    fn test_tokenize_string() {
        let string = "'qwertyuiop \"world\" 1234567890!'".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::String,
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0,
                    character_index_start: 0,
                    character_index_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_identifier() {
        let string = "Hello_world_123".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::Identifier,
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0,
                    character_index_start: 0,
                    character_index_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_keyword() {
        assert_eq!(
            tokenize("let type".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::KeywordLet,
                    representation: "let".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 0,
                        column_end: 2,
                        character_index_start: 0,
                        character_index_end: 0
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 3,
                        column_end: 3,
                        character_index_start: 0,
                        character_index_end: 0
                    }
                },
                Token {
                    token_type: TokenType::KeywordType,
                    representation: "type".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 4,
                        column_end: 7,
                        character_index_start: 0,
                        character_index_end: 0
                    }
                }
            ])
        )
    }
}

#[cfg(test)]
mod test_transpile {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_let_statement_1() {
        assert_eq!(
            transpile_source("let x = '123'".to_string()),
            Ok("const x = '123'".to_string())
        )
    }

    #[test]
    fn test_let_statement_2() {
        assert_eq!(
            transpile_source("let x = y".to_string()),
            Ok("const x = y".to_string())
        )
    }

    #[test]
    fn test_number_literal() {
        assert_debug_snapshot!(transpile_source("let x = -123.456".to_string()))
    }

    #[test]
    fn test_array_literal_1() {
        assert_debug_snapshot!(transpile_source("let x = [[],[2],[3,4,],]".to_string()))
    }

    #[test]
    fn test_tag_literal_1() {
        assert_debug_snapshot!(transpile_source("let x = #ok(#some(2))".to_string()))
    }

    #[test]
    fn test_function_1() {
        assert_debug_snapshot!(transpile_source("let x = \\x => 'yo'".to_string()))
    }
    #[test]
    fn test_function_2() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#red => 'red' \\ #blue => 'blue'".to_string()
        ))
    }

    #[test]
    fn test_function_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#ok(#some(r)) => r \\ _ => 'yo'".to_string()
        ))
    }

    #[test]
    fn test_function_4() {
        assert_debug_snapshot!(transpile_source(
            "let or = \\(#false, #false,) => #false \\(_, _) => #true".to_string()
        ))
    }

    #[test]
    fn test_function_5() {
        assert_debug_snapshot!(transpile_source(
            "let square = \\(x: number): number => x".to_string()
        ))
    }

    #[test]
    fn test_function_call_1() {
        assert_debug_snapshot!(transpile_source("let x = y.square.bomb".to_string()))
    }

    #[test]
    fn test_function_call_2() {
        assert_debug_snapshot!(transpile_source("let x = a.times(b).and(c, d)".to_string()))
    }

    #[test]
    fn test_function_call_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = x.(\\#ok(x) => a \\#no => b)".to_string()
        ))
    }

    #[test]
    fn test_record_1() {
        assert_debug_snapshot!(transpile_source(
            "let x = {a: int = {b = c}, d = e}".to_string()
        ))
    }

    #[test]
    fn test_record_2() {
        assert_debug_snapshot!(transpile_source(
            "let x = {...{x = 2, y = 3}, y = 4, z = 5}".to_string()
        ))
    }

    #[test]
    fn test_block_1() {
        assert_debug_snapshot!(transpile_source(
            "let x = let x = a let y = b 'hello world'".to_string()
        ))
    }

    #[test]
    fn test_monadic_binding_1() {
        assert_debug_snapshot!(transpile_source(
            "
            let f = \\(a, b) => 
                let #some(x) = a
                let #ok(y) = b 
                x.plus(y)
            "
            .to_string()
        ))
    }

    #[test]
    fn test_monadic_binding_2() {
        assert_debug_snapshot!(transpile_source(
            "
            let f = \\(a, b) => 
                let #some(x) = a
                let #ok(y) = b 
                    else \\_ => #none
                x.plus(y)
            "
            .to_string()
        ))
    }

    #[test]
    fn test_destructure_record_1() {
        assert_debug_snapshot!(transpile_source(
            "let f = \\{a, b= #ok(#some(c))} => a.plus(c)".to_string()
        ))
    }

    #[test]
    fn test_type_annotation_1() {
        assert_debug_snapshot!(transpile_source("let x: number = 1".to_string()))
    }

    #[test]
    fn test_type_alias_1() {
        assert_debug_snapshot!(transpile_source(
            "type Shape = #circle({radius: string}) | #none | #square(number)".to_string()
        ))
    }

    #[test]
    fn transpile_let_destructure_number_1() {
        assert_debug_snapshot!(transpile_source("let f = \\x => let 2 = x 10".to_string()))
    }

    #[test]
    fn transpile_let_destructure_string_1() {
        assert_debug_snapshot!(transpile_source(
            "let f = \\x => let 'yo' = x 'hey'".to_string()
        ))
    }

    #[test]
    fn transpiled_destructured_array_7() {
        // expected error, cannot have more than one spread
        assert_debug_snapshot!(transpile_source(
            "let arrayOnlyOneElement = 
                \\[a,...b,...c,d] => true
                \\_ => false
            "
            .to_string()
        ))
    }
}

#[cfg(test)]
mod test_type_check {

    // #[test]
    // fn test_generic_function_with_specified_type_variable() {
    //     assert_debug_snapshot!(type_check_source(
    //         "
    //         let as<T> = \\(x: T): T => x
    //         let x = 'hello'.as<number>
    //         "
    //         .trim()
    //         .to_string()
    //     ))
    // }
    //
}
