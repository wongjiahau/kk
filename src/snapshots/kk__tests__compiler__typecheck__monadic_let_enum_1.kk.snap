---
source: src/main.rs
expression: stripped_actual_output.trim()
---
============
INPUT
============
enum Color = #red #green
enum Boolean = #true #false
let firstIsTrue = \x =>
  let #red = x
  #true

============
EXIT CODE
============
1

============
STDOUT
============


============
STDERR
============
error:
  ┌─ tests/compiler/typecheck/monadic_let_enum_1.kk:5:3
  │
5 │   #true
  │   ^^^^^ Type mismatch
  │
  = +---------------+-------------+
    | EXPECTED TYPE | ACTUAL TYPE |
    +---------------+-------------+
    | - Color       | + Boolean   |
    +---------------+-------------+
